Updates to ZIP 32 and ZIP 224 for Orchard (#448)
zcash/zips: zip-0032.rst

- Use :math:`I_R` as the child chain code :math:`\mathsf{c}_i`.

Specification: Orchard key derivation
=====================================

The derivation mechanism for Sapling addresses specified above incurs significant complexity to support
non-hardened derivation. In the several years since Sapling was deployed, we have seen no use cases for
non-hardened derivation appear. With that in mind, we define Orchard key derivation very similarly to
Sprout above: only hardened derivation is supported.

Orchard extended keys
---------------------

We represent an Orchard extended spending key as :math:`(\mathsf{sk, c}),` where :math:`\mathsf{sk}`
is the normal Orchard spending key (opaque 32 bytes), and :math:`\mathsf{c}` is the chain code.

Orchard master key generation
-----------------------------

Let :math:`S` be a seed byte sequence of a chosen length, which MUST be at least 32 and at most 252 bytes.

- Calculate :math:`I = \mathsf{BLAKE2b}\text{-}\mathsf{512}(\texttt{“ZcashIP32Orchard”}, S)`.
- Split :math:`I` into two 32-byte sequences, :math:`I_L` and :math:`I_R`.
- Use :math:`I_L` as the master spending key :math:`\mathsf{sk}_m`.

> - Use :math:`I_R` as the master chain code :math:`\mathsf{c}_m`.

Reviewer: ```suggestion
- Use :math:`I_R` as the master chain code :math:`\mathsf{c}_m`.
- Return :math:`(\mathsf{sk}_m, \mathsf{c}_m)` as the master extended spending key :math:`m_\mathsf{Orchard}`.
```

    :math:`I = \mathsf{PRF^{expand}}(\mathsf{c}_{par}, [\texttt{0x81}]\,||\,\mathsf{sk}_{par}\,||\,\mathsf{I2LEOSP}_{32}(i))`.
  - If not (normal child): return failure.

- Split :math:`I` into two 32-byte sequences, :math:`I_L` and :math:`I_R`.
- Use :math:`I_L` as the child spending key :math:`\mathsf{sk}_{i}`.
- Use :math:`I_R` as the child chain code :math:`\mathsf{c}_i`.

Orchard diversifier derivation
------------------------------

As with Sapling, we define a mechanism for deterministically deriving a sequence of diversifiers, without
leaking how many diversified addresses have already been generated for an account. Unlike Sapling, we do so
by deriving a diversifier key directly from the full viewing key, instead of as part of the extended spending
key. This means that the full viewing key provides the capability to determine the position of a diversifier
within the sequence, which matches the capabilities of a Sapling extended full viewing key but simplifies the
key structure.

Given an Orchard extended spending key :math:`(\mathsf{sk}_i, \mathsf{c}_i)`:

- Let :math:`\mathsf{fvk}_i` be the raw encoding of the Orchard full viewing key for :math:`\mathsf{sk}_i`

>   (as specified in TODO).

Reviewer: ```suggestion
  (as specified in #[protocol-orchardfullviewingkeyencoding]_).
```

- Use :math:`I_R` as the child chain code :math:`\mathsf{c}_i`.

Specification: Orchard key derivation
=====================================

The derivation mechanism for Sapling addresses specified above incurs significant complexity to support
non-hardened derivation. In the several years since Sapling was deployed, we have seen no use cases for
non-hardened derivation appear. With that in mind, we define Orchard key derivation very similarly to
Sprout above: only hardened derivation is supported.

Orchard extended keys
---------------------

We represent an Orchard extended spending key as :math:`(\mathsf{sk, c}),` where :math:`\mathsf{sk}`

> is the normal Orchard spending key (opaque 32 bytes), and :math:`\mathsf{c}` is the chain code.

Reviewer 2: Chain code is not defined in the terminology of this document yet, could you add it?

Orchard master key generation
-----------------------------

Let :math:`S` be a seed byte sequence of a chosen length, which MUST be at least 32 and at most 252 bytes.

- Calculate :math:`I = \mathsf{BLAKE2b}\text{-}\mathsf{512}(\texttt{“ZcashIP32Orchard”}, S)`.
- Split :math:`I` into two 32-byte sequences, :math:`I_L` and :math:`I_R`.
- Use :math:`I_L` as the master spending key :math:`\mathsf{sk}_m`.
- Use :math:`I_R` as the master chain code :math:`\mathsf{c}_m`.
- Return :math:`(\mathsf{sk}_m, \mathsf{c}_m)` as the master extended spending key
  :math:`m_\mathsf{Orchard}`.

Orchard child key derivation
----------------------------

:math:`\mathsf{CDKsk}((\mathsf{sk}_{par}, \mathsf{c}_{par}), i)`:math:`\rightarrow (\mathsf{sk}_{i}, \mathsf{c}_i)`

- Check whether :math:`i \geq 2^{31}` (whether the child is a hardened key).

  - If so (hardened child): let
    :math:`I = \mathsf{PRF^{expand}}(\mathsf{c}_{par}, [\texttt{0x81}]\,||\,\mathsf{sk}_{par}\,||\,\mathsf{I2LEOSP}_{32}(i))`.
  - If not (normal child): return failure.

- Split :math:`I` into two 32-byte sequences, :math:`I_L` and :math:`I_R`.
- Use :math:`I_L` as the child spending key :math:`\mathsf{sk}_{i}`.

> - Use :math:`I_R` as the child chain code :math:`\mathsf{c}_i`.

Reviewer 3: It would be helpful to add "chain code" to the terminology section. @daira explained it as a capability for child key derivation, which makes more sense to me than the BIP 32 definition which specifies it as "additional randomness" but doesn't give a justification for needing that extra entropy.

Orchard diversifier derivation
------------------------------

As with Sapling, we define a mechanism for deterministically deriving a sequence of diversifiers, without
leaking how many diversified addresses have already been generated for an account. Unlike Sapling, we do so

> by deriving a diversifier key directly from the full viewing key, instead of as part of the extended spending

Reviewer 2: Line wrap?

Reviewer: This isn't longer than some other paragraph lines in the document.

key. This means that the full viewing key provides the capability to determine the position of a diversifier
within the sequence, which matches the capabilities of a Sapling extended full viewing key but simplifies the
key structure.

Given an Orchard extended spending key :math:`(\mathsf{sk}_i, \mathsf{c}_i)`:

- Let :math:`\mathit{FVK}_i` be the raw encoding of the Orchard full viewing key for :math:`\mathsf{sk}_i`
  (as specified in [#spec-encoding-orchard-fvk]_).

> - :math:`\mathsf{dk}_i = \mathsf{truncate}_{32}(\mathsf{PRF^{expand}}(\mathit{FVK}_i, [\texttt{0x82}]))`.

Reviewer 3: Where does this 0x82 come from?

Author: It's a new tag, within the set of tags used with `PRF^expand` in the Zcash protocol. I picked it somewhat arbitrarily.

- Let :math:`j` be the index of the desired diversifier, in the range :math:`0\,.\!. 2^{88} - 1`.
- :math:`d_{i,j} = \mathsf{FF1}\text{-}\mathsf{AES256.Encrypt}(\mathsf{dk}_i, \texttt{“”}, \mathsf{I2LEBSP}_{88}(j))`.

Note that unlike Sapling, all Orchard diversifiers are valid, and thus all possible values of :math:`j` yield
valid diversifiers.

> The default diversifier for :math:`(\mathsf{sk}_i, \mathsf{c}_i)` is defined to be :math:`d_{i,0}.`

Reviewer 2: 👍

* :math:`m_\mathsf{Sprout} / purpose' / coin\_type' / account' / address\_index`.

Orchard key path
----------------

Orchard supports diversified addresses with the same spending authority (like Sapling). A group of such
addresses shares the same full viewing key and incoming viewing key, and so creating as many unlinkable
addresses as needed does not increase the cost of scanning the block chain for relevant transactions.

The above key path levels include an account identifier, which in all user interfaces is represented as a
"bucket of funds" under the control of a single spending authority. Therefore, wallets implementing Orchard
ZIP 32 derivation MUST support the following path for any account in range :math:`\{ 0\,.\!. 2^{31} - 1 \}`:

* :math:`m_\mathsf{Orchard} / purpose' / coin\_type' / account'`.

Furthermore, wallets MUST support generating the default payment address (corresponding to the default
diversifier for Orchard) for any account they support. They MAY also support generating a stream of payment
addresses for a given account, if they wish to maintain the user experience of giving a unique address to

> each recipient.

Reviewer: The motivation for this is primarily security, not maintaining user experience. See the note at the end of https://zips.z.cash/protocol/protocol.pdf#addressesandkeys .

intended for optimizing performance of address lookups, and MUST NOT be assumed to uniquely identify
a particular address.

Orchard Full Viewing Key Fingerprints and Tags
----------------------------------------------

An "Orchard full viewing key fingerprint" of a full viewing key with raw encoding :math:`FVK` (as specified

> in TODO) is given by:

Reviewer: ```suggestion
in #[spec-encoding-fvk]_) is given by:
```

> An "Orchard full viewing key fingerprint" of a full viewing key with raw encoding :math:`FVK` (as specified
> in TODO) is given by:

Reviewer: ```suggestion
An "Orchard full viewing key fingerprint" of a full viewing key with raw encoding :math:`\mathit{FVK}`
(as specified in [#spec-encoding-fvk]_) is given by:
```

> * :math:`\mathsf{BLAKE2b}\text{-}\mathsf{256}(\texttt{“ZcashOrchardFVFP”}, FVK)`.

Reviewer: ```suggestion
* :math:`\mathsf{BLAKE2b}\text{-}\mathsf{256}(\texttt{“ZcashOrchardFVFP”}, \mathit{FVK})`.
```

are encoded using Bech32 even though other Sprout keys and addresses are encoded
using Base58Check.

Orchard extended spending keys
------------------------------

An Orchard extended spending key :math:`(\mathsf{sk, c})`, at depth :math:`depth`, with parent full viewing
key tag :math:`parent\_fvk\_tag` and child number :math:`i`, is represented as a byte sequence:

* :math:`\mathsf{I2LEOSP}_{8}(depth)\,||\,parent\_fvk\_tag\,||\,\mathsf{I2LEOSP}_{32}(i)\,||\,\mathsf{c}\,||\,\mathsf{sk}`.

For the master extended spending key, :math:`depth` is :math:`0`, :math:`parent\_fvk\_tag` is
4 zero bytes, and :math:`i` is :math:`0`.

When encoded as Bech32, the Human-Readable Part is ``secret-orchard-extsk-main``
for the production network, or ``secret-orchard-extsk-test`` for the test network.

We define this encoding for completeness, however given that it includes the capability to derive child
spending keys, we expect that most wallets will only expose the regular Orchard spending key encoding to

> users (TODO: reference).

Reviewer: ```suggestion
users [#spec-encoding-sk]_.
```

are encoded using Bech32 even though other Sprout keys and addresses are encoded
using Base58Check.

Orchard extended spending keys
------------------------------

An Orchard extended spending key :math:`(\mathsf{sk, c})`, at depth :math:`depth`, with parent full viewing
key tag :math:`parent\_fvk\_tag` and child number :math:`i`, is represented as a byte sequence:

* :math:`\mathsf{I2LEOSP}_{8}(depth)\,||\,parent\_fvk\_tag\,||\,\mathsf{I2LEOSP}_{32}(i)\,||\,\mathsf{c}\,||\,\mathsf{sk}`.

For the master extended spending key, :math:`depth` is :math:`0`, :math:`parent\_fvk\_tag` is
4 zero bytes, and :math:`i` is :math:`0`.

When encoded as Bech32, the Human-Readable Part is ``secret-orchard-extsk-main``

> for the production network, or ``secret-orchard-extsk-test`` for the test network.

Reviewer 2: ```suggestion
for Mainnet, or ``secret-orchard-extsk-test`` for Testnet.
```
